From 133d16f2c0bfec8e02e416a96958b42a509e9d9b Mon Sep 17 00:00:00 2001
From: Vladimir Barinov <vladimir.barinov@cogentembedded.com>
Date: Sat, 2 Jan 2021 01:09:56 +0300
Subject: [PATCH] media: rcar_csi2: fix VC cotrnol

The VCDT is not updated until hwinit is called.
Fix it.

Singed-off-by: Vladimir Barinov <vladimir.barinov@cogentembedded.com>
---
 drivers/media/platform/soc_camera/rcar_csi2.c | 23 ++++++++++++++---------
 1 file changed, 14 insertions(+), 9 deletions(-)

diff --git a/drivers/media/platform/soc_camera/rcar_csi2.c b/drivers/media/platform/soc_camera/rcar_csi2.c
index 73ef28f..3154536 100644
--- a/drivers/media/platform/soc_camera/rcar_csi2.c
+++ b/drivers/media/platform/soc_camera/rcar_csi2.c
@@ -680,15 +680,6 @@ static int rcar_csi2_s_power(struct v4l2_subdev *sd, int on)
 
 	mutex_lock(&priv->lock);
 
-	if (idx < 2) {
-		priv->vcdt &= ~(RCAR_CSI2_VCDT_VCDTN_EN << (idx * 16));
-		priv->vcdt |= (on ? RCAR_CSI2_VCDT_VCDTN_EN << (idx * 16) : 0);
-	} else {
-		idx -= 2;
-		priv->vcdt2 &= ~(RCAR_CSI2_VCDT_VCDTN_EN << (idx * 16));
-		priv->vcdt2 |= (on ? RCAR_CSI2_VCDT_VCDTN_EN << (idx * 16) : 0);
-	}
-
 	if (on) {
 		if (atomic_inc_return(&priv->use_count) == 1) {
 			pm_runtime_get_sync(&priv->pdev->dev);
@@ -696,7 +687,21 @@ static int rcar_csi2_s_power(struct v4l2_subdev *sd, int on)
 			if (ret < 0)
 				return ret;
 		}
+
+		if (idx < 2) {
+			priv->vcdt |= (RCAR_CSI2_VCDT_VCDTN_EN << (idx * 16));
+		} else {
+			idx -= 2;
+			priv->vcdt2 |= (RCAR_CSI2_VCDT_VCDTN_EN << (idx * 16));
+		}
 	} else {
+		if (idx < 2) {
+			priv->vcdt &= ~(RCAR_CSI2_VCDT_VCDTN_EN << (idx * 16));
+		} else {
+			idx -= 2;
+			priv->vcdt2 &= ~(RCAR_CSI2_VCDT_VCDTN_EN << (idx * 16));
+		}
+
 		if (atomic_dec_return(&priv->use_count) == 0) {
 			rcar_sci2_debug_show(priv);
 			rcar_csi2_hwdeinit(priv);
-- 
2.7.4

